Drag snapshot does not paint transformed content

nullhook opened the pull request:
#### 860e78e7dd0ab13332da3636b1bf3f2e499344f7
<pre>
Drag snapshot does not paint transformed content
<a href="https://bugs.webkit.org/show_bug.cgi?id=267811">https://bugs.webkit.org/show_bug.cgi?id=267811</a>
<a href="https://rdar.apple.com/121382647">rdar://121382647</a>

Reviewed by Simon Fraser.

Previously, when taking a snapshot of a node,
any applied transforms were ignored. This caused
transformed elements to be missing in the snapshot image.

This fix ensures that node snapshots now correctly
account for transforms.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::snapshotNode):

Use logical size as ImageData understands
logical size, not backend size.

Canonical link: <a href="https://commits.webkit.org/296395@main">https://commits.webkit.org/296395@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23d43f0dd97573a5ff994f46df5d6d7daf78c901

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108352 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18436 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113561 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58781 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110315 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36563 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82278 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111300 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22758 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97600 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62713 "Found 1 new API test failure: /WPE/TestDownloads:/webkit/Downloads/mime-type (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22175 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15737 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58288 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92126 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15793 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116682 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26080 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91301 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35779 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93876 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91102 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23216 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35988 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13758 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31162 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35308 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40846 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35023 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38376 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36703 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3743,7 +3743,7 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
     }();
 
-    if (shouldPaintContent && paintingInfo.subtreePaintRoot) {
+    if (shouldPaintContent && paintingInfo.subtreePaintRoot && paintingInfo.subtreePaintRoot->hasLayer()) {
```
Please explain this change in the commit message. Adding this check will change behavior for `position:static` subtree roots with `position:absolute` descendants.

nullhook replied:
drag snapshots didn’t work for normal flow elements because `enclosingLayer()` found the nearest layer, which in the case below was the HTML root layer. as a result, the layer being used for painting decisions was a much higher ancestor, not the specified div itself. my intent here was to use `hasLayer()` in conjunction with `enclosingLayer()` so that it finds the specified draggable div’s layer itself, rather than another layer.

basically, the conditional logic should apply only to elements with their own layers, and the above code is meant to guard that

```
<div draggable="true">Drag me</div>
```

nullhook replied:
will address this in a separate PR

smfr commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -7926,3 +7926,5 @@ webkit.org/b/294274 [ Debug ] platform/ios/mediastream/audio-muted-in-background
 
 fast/snapshot/nested-stacking-context.html [ Skip ]
 fast/snapshot/positioned-abs-relative-body.html [ Skip ]
+fast/snapshot/positioned-static.html [ Skip ]
```
Should you just skip the whole directory?

smfr commented on LayoutTests/fast/snapshot/positioned-static.html:
```diff
@@ -0,0 +1,55 @@
+</style>
+</head>
+<body>
+    <div id="box">Is this included?</div>
```
Please include the testa case with a static position div having an abspos child.

nullhook replied:
will address this in a separate PR

smfr commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -911,7 +911,7 @@ void RenderObject::addAbsoluteRectForLayer(LayoutRect& result)
 LayoutRect RenderObject::paintingRootRect(LayoutRect& topLevelRect)
 {
-    LayoutRect result = absoluteBoundingBoxRectIgnoringTransforms();
+    LayoutRect result = absoluteBoundingBoxRect(true);
```
`true` is the default; no need to specify it.

smfr reviewed (approved):
Please rewrap the commit message to 80 columns.

nullhook committed:
Drag snapshot does not paint transformed content
https://bugs.webkit.org/show_bug.cgi?id=267811
rdar://121382647

Reviewed by Simon Fraser.

Previously, when taking a snapshot of a node,
any applied transforms were ignored. This caused
transformed elements to be missing in the snapshot image.

This fix ensures that node snapshots now correctly
account for transforms.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::snapshotNode):

Use logical size as ImageData understands
logical size, not backend size.

Canonical link: https://commits.webkit.org/296395@main
